streams/shared/src/main/scala/zio/stream/experimental/ZTransducer.scala

```diff
@@ -124,6 +128,223 @@ object ZTransducer {
   def fromPush[R, E, I, O](push: Option[Chunk[I]] => ZIO[R, E, Chunk[O]]): ZTransducer[R, E, I, O] =
     ZTransducer(Managed.succeed(push))
 
+  /**
+   * Creates a transducer by folding over a structure of type `S`.
+   */
+  def fold[I, O](z: O)(contFn: O => Boolean)(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    ZTransducer {
+      final case class FoldState(started: Boolean, result: O)
+
+      val initial = FoldState(false, z)
+
+      @tailrec def go(in: Chunk[I], os0: Chunk[O], state: FoldState): (Chunk[O], FoldState) =
+        in.headOption match {
+          case None => os0 -> state
+          case Some(i) =>
+            val o = f(state.result, i)
+            if (contFn(o))
+              go(in.drop(1), os0, FoldState(true, o))
+            else
+              go(in.drop(1), os0 + o, initial)
+        }
+
+      ZRef.makeManaged(initial).map { state =>
+        {
+          case Some(in) => state.modify(go(in, Chunk.empty, _))
+          case None     => state.getAndSet(initial).map(s => if (s.started) Chunk.single(s.result) else Chunk.empty)
```

Comment:
Yes, though when it comes to responding to `push(None)`, you have to account for the possibility that the transducer is in the middle of a fold, because if it isn't you cannot produce `Chunk(z)` but `Chunk.empty`.

Reply:
You can always emit `Chunk(state.z)` if the initial `FoldState` contains the `z` passed to the function, no?

Reply:
So it's fine the transducer outputs `Chunk(z)` ad infinitum?

Reply:
I guess that makes sense, as that is the behaviour on an empty input. Thanks, will do that!

Reply:
Not so sure after all. I updated foldDecompose just for an example, but it breaks foldUntil because now when the stream pushes None the zero will be output and there is no way to know whether it is a valid output calculated from some previous runs, or it is because it was reset already.

Reply:
Can you write out the inputs for this case?

Reply:
Yep, the test:
```scala
ZStream[Long](1, 1, 1, 1, 1, 1)
  .aggregate(ZTransducer.foldUntil(0L, 3)(_ + (_: Long)))
  .runCollect
```
yields `List(3,3,0)` instead of `List(3,3)`

Reply:
Cool. That's a great example. I agree that the result is not satisfactory in this case.

For `ZStream.empty` though, it still seems that we should emit the initial state. What do you think? I'll check what other libraries are doing in this case.

Reply:
We should also make sure that we're consistent with `ZStream#mapAccum` as a transducer fold is basically a scan.

Reply:
> For ZStream.empty though, it still seems that we should emit the initial state. What do you think? I'll check what other libraries are doing in this case.

I agree. I can just replace the current state with `Option[O]`, set it to `Some(z)` initially and reset to `None` after each emit.

Reply:
@regiskuckaertz Sounds perfect.